### DISCUSSION TITLE
Redirecting to Specific IDP Forms using OIDC Plugin

### DISCUSSION THREAD

**Discussion by MarinCouderc** [Bdaya-Dev/oidc#54] [Q&A] [answered]
Hello,

We are developing a Flutter application and are using Keycloak as our identity provider (IDP). We have successfully implemented the oidc plugin for authentication purposes, and it has been working seamlessly for our login process.

However, we have encountered a challenge that we hope you can help us with. Our application features a native Flutter homepage that includes two buttons: one for registration and another for login. Initially, we were able to direct users to the appropriate Keycloak form based on their selection—either to the registration form or the login form.

We understand that this requirement might not be related to Dynamic Client Registration. However, we believe there should be a way to specify the desired IDP URI to the oidc plugin to direct users to the correct Keycloak page based on their choice (for example as a parameter to loginAuthorizationCodeFlow). We are reaching out to inquire if this use case is supported by the oidc plugin, and if so, could you please provide guidance or insights on how to achieve this?

Thanks!

**Comment by ahmednfwela** [comment]
Hello, 
You can pass arbitrary query parameters using the `extraParameters` field when authenticating.

https://github.com/Bdaya-Dev/oidc/blob/4e9fbaca58a0c86abcb592c77d06b38706b178b5/packages/oidc/lib/src/managers/user_manager.dart#L117

does this solve your use-case @MarinCouderc ?

**Reply by MarinCouderc** [reply]
Hello,
Unfortunately not (at least from my understanding), as it looks like Keycloak doesn't have any query parameters that could be used to redirect to the sign-up page, the difference being in the path itself (/auth vs /registrations)

**Reply by ahmednfwela** [reply]
I see, so you want to override the `authorization_endpoint` from the discovery document to the sign-up url offered by keycloak ?
but then that's not spec-compliant, and it will not work since keycloak won't expect it.

I suggest merging the "sign up" and "login" buttons into a single "authorize" button to avoid the issue altogether.

**Reply by MarinCouderc** [reply]
Kind of. What would help us the most would be similar to what openid_client package does, eg. the login method would give us a Function(String) parameter we could use to get the current uri, and when needed replace '/auth' by '/registrations' (which is what we're doing now, but would rather migrate to oidc plugin as it is way more stable). Hope what I'm saying makes sense!

**Reply by ahmednfwela** [reply]
I can introduce an authorization url override parameter to work around this

**Reply by MarinCouderc** [reply]
That would be perfect, thanks!

**Reply by ahmednfwela** [reply, marked as answer]
@MarinCouderc Just published a new version that addresses this, waiting for your feedback!

**Reply by ahmednfwela** [reply]
Relevant docs entry: https://bdaya-dev.github.io/oidc/oidc-usage/#overriding-the-discovery-document

**Reply by MarinCouderc** [reply]
@ahmednfwela Tested it, it works perfectly well, thanks a lot for your very quick support!

**Reply by s681562** [reply]
@MarinCouderc  Could you please share the code / authorizationEndpointUri?

I am using this for direct going to Register form. But doesn't work. I got missing cookie message.

Uri authorizationEndpointUri = Uri.parse("https://.../realms/.../**login-actions/registration**"); await manager!.loginAuthorizationCodeFlow( discoveryDocumentOverride: manager!.discoveryDocument.copyWith( authorizationEndpoint: authorizationEndpointUri, ) );

I am using this for direct going to Forgot Password form. But when click on Back to Login, I got missing cookie message, instead.

Uri authorizationEndpointUri = Uri.parse("https://.../realms/.../**login-actions/reset-credentials**"); await manager!.loginAuthorizationCodeFlow( discoveryDocumentOverride: manager!.discoveryDocument.copyWith( authorizationEndpoint: authorizationEndpointUri, ) );

Which params are missing in loginAuthorizationCodeFlow?

**Reply by ahmednfwela** [reply]
hi @s681562 can you please elaborate on what error message you got? was it from keycloak or from the client side ?

**Reply by s681562** [reply]
When I am using these authorizationEndpointUri, I got message from keycloak, that some cookies are not found. 
That's all. Action stopped.
So i guess, some settings are missing inside oidcmanager.

**Reply by ahmednfwela** [reply]
@s681562 I doubt it's a problem with the plugin, since you didn't get an oidc related error, meaning login was actually successful.

I believe this is solved from keycloack itself, check this [SO](https://stackoverflow.com/questions/76760999/why-after-login-keycloak-display-cookie-not-found-please-make-sure-cookies-are)
